src/main/kotlin/org/jitsi/rtp/rtp/RtpPacket.kt

```diff
@@ -242,136 +308,81 @@ open class RtpPacket(
-        repeat(numPaddingBytes) {
-            // Set the padding to 0 (we have to do this because we may be
-            // reusing a buffer).
-            newBuffer[newHeaderLength++] = 0
+        if (buffer !== newBuffer || offset != newOffset) {
+            // Copy the base header into place.
+            System.arraycopy(buffer, offset, newBuffer, newOffset, baseHeaderLength)
         }
 
-        newBuffer.putShort(extHeaderOffset + 2, ((extensionBytes + numPaddingBytes) / 4).toShort())
-
-        // Now we have the new header, with the added header extension and with
-        // the correct padding, in newBuffer at offset 0. Lets move it to the
-        // correct place (right before the payload).
-        val newOffset = newPayloadOffset - newHeaderLength
-        if (newOffset != 0) {
-            System.arraycopy(newBuffer, 0, newBuffer, newOffset, newHeaderLength)
+        if (pendingHeaderExtensions.isNotEmpty()) {
+            var off = newOffset + baseHeaderLength
+            // Write the header extension
+            newBuffer.putShort(off, 0xBEDE.toShort())
+            newBuffer.putShort(off + 2, (newExtHeaderLength / 4).toShort())
+            off += 4
+            // Write pending header extension elements
+            pendingHeaderExtensions.forEach { h ->
+                System.arraycopy(h.currExtBuffer, h.currExtOffset, newBuffer, off, h.currExtLength)
+                off += h.currExtLength
+            }
+            // Write padding
+            while (off < newOffset + newHeaderLength) {
+                newBuffer[off] = 0
+                off++
+            }
         }
 
-        // All that is left to do is update the NewRawPacket state.
         val oldBuffer = buffer
         buffer = newBuffer
         // Reference comparison to see if we got a new buffer.  If so, return the old one to the pool
         if (oldBuffer !== newBuffer) {
             BufferPool.returnArray(oldBuffer)
         }
         offset = newOffset
-        length = newHeaderLength + currPayloadLength
+        length = newPacketLength
+        headerLength = newHeaderLength
 
         // ... and set the extension bit.
-        hasExtensions = true
-
-        // Setup the single HeaderExtension instance of this NewRawPacket and
-        // return it.
-        val newExt = headerExtensions.currHeaderExtension
-        newExt.setOffsetLength(offset + extensionDataOffset, extDataLength + 1)
-
-        // Update the header length
-        headerLength = RtpHeader.getTotalLength(buffer, offset)
+        hasEncodedExtensions = pendingHeaderExtensions.isNotEmpty()
 
-        return newExt
+        // Clear pending extensions.
+        this.pendingHeaderExtensions = null
     }
 
     /**
      * Return the total length of the extensions in this packet, including the extension header
      */
     private val extensionBlockLength: Int
         get() {
-            if (!hasExtensions) {
+            if (!hasEncodedExtensions) {
```

Comment:
It might be cleaner to inline this function (it's only used once and doesn't need the `hasEncodedExtensions` check).